install.sh

```diff
@@ -276,77 +279,171 @@ assert_usable_kernel() {
+  if [ ! -e /etc/sysctl.conf ] ; then
+    SYSCTL_PROBABLY_WORKS="no"
+  fi
 
-  return 0;
+  # OK, so they need the sysctl set.
+  #
+  # If, however, they _can't_ set the sysctl, make the install fail
+  # now.
+  if [ "$SYSCTL_PROBABLY_WORKS" = "no" ] ; then
+    echo "# sysctl -w kernel.unprivileged_userns_clone=1"
+    echo "# cat >> /etc/sysctl.conf << __EOF__"
+    echo ""
+    echo "# Enable non-root users to create sandboxes (needed by Sandstorm)."
+    echo "kernel.unprivileged_userns_clone = 1"
+    echo "__EOF__"
+    echo ""
+    fail "You are using a Debian-derived Linux kernel, which needs a configuration option" \
+         "set in order to run Sandstorm. To set that option, please run the shell commands" \
+         "above as root. (If you are running this in a container through e.g. Docker, you will" \
+         "have to run the above commands _outside_ the container.)"
+  fi
+
+  # OK, so either it already works, or we know it's a reasonable idea to ask
+  # the user to let us enable the sysctl.
 }
-__EOF__
-    if cc /tmp/sandstorm-userns-test.c -o /tmp/sandstorm-userns-test; then
-      if ! /tmp/sandstorm-userns-test; then
-        rm -f /tmp/sandstorm-userns-test /tmp/sandstorm-userns-test.c
-        fail "Your kernel does not appear to be compiled with" \
-             "support for unprivileged user namespaces (CONFIG_USER_NS=y), or something else is" \
-             "preventing creation of user namespaces. This feature is critical for sandboxing." \
-             "Arch Linux is known to ship with a kernel that disables this feature; if you are" \
-             "using Arch, you will unfortunately need to compile your own kernel (see" \
-             "https://bugs.archlinux.org/task/36969). If you are not using Arch, and don't" \
-             "know why your system wouldn't have user namespaces, please file a bug against" \
-             "Sandstorm so we can figure out what happened."
-      fi
-    else
-      echo "WARNING: Couldn't compile user namespace test. We'll assume user namespaces" >&2
-      echo "  are enabled." >&2
+
+enable_userns_sysctl_if_needed() {
+  # This function enables the Debian/Ubuntu-specific unprivileged
+  # userns sysctl.
+  #
+  # We only run it if we need to.
+  if [ "$USERNS_CLONE_UNPRIVILEGED_NEEDS_SYSCTL_SET" != "yes" ] ; then
+    return
+  fi
+
+  # It only makes sense when running as root, so if we are not
+  # currently running as root, just skip this code.
+  if [ "no" = "$CURRENTLY_UID_ZERO" ] ; then
+    return
+  fi
+
+  local PRINT_USERNS_INFO="yes"
+  local ACCEPTED_SYSCTL_SWITCH="no"
+
+  if [ "yes" = "${ACCEPTED_FULL_SERVER_INSTALL:-}" ] ; then
+    PRINT_USERNS_PROMPT="no"
+    ACCEPTED_SYSCTL_SWITCH="yes"
+  fi
+
+  if [ "${PRINT_USERNS_PROMPT}" = "yes" ] ; then
+    echo "Sandstorm requires sysctl kernel.unprivileged_userns_clone to be enabled."
+    echo "Currently, it is not enabled on your system."
+    if prompt-yesno "Shall I enable it for you?" yes; then
```

Comment:
Should we give some indication that we plan to enable it permanently? (Mainly so that the rare user who has any idea what a sysctl is doesn't get the impression that they'll have to go do that themselves later.)